### PR TITLE
Add judge model, support simple qa and chinese simple qa

### LINKED ISSUE #381: hellaswag 相关问题
## 使用的工具 / Tools Used
- [x] Native / 原生框架
- [ ] Opencompass backend
- [ ] VLMEvalKit backend
- [ ] RAGEval backend
- [ ] Perf / 模型推理压测工具
- [ ] Arena  /竞技场模式

## 问题描述 / Issue Description

最近在基于 api 的评测（`eval_type=EvalType.SERVICE`）中发现模型在 hellaswag 上跑不出结果且耗时非常长，预测的数据基本都答非所问、比较冗长，对比了基于 OpenCompass 后端在 hellaswag 上进行评测的结果，发现 EvalScope 原生支持的 hellaswag 中没有把它当成一个多选任务：
```python
# benchmarks/hellaswag/hellaswag_adapter.py
    model_adapter=OutputType.CONTINUOUS,
    output_types=[OutputType.CONTINUOUS, OutputType.GENERATION],
```
而是跟 truthful_qa 一样作为了基于 logits 的评测，但在文档中（https://evalscope.readthedocs.io/zh-cn/latest/get_started/supported_dataset.html#id2 ）没有提及，文档中只注明了 truthful_qa 这一项。

另外，注册时的 `prompt_template` 也没有用到：
```python
# benchmarks/hellaswag/hellaswag_adapter.py
    prompt_template=
    'Respond with the index of sentence that makes the most sense, chose from 0, 1, 2, 3, derive your final answer as `The answer is ...`.',  # noqa: E501
```

想问一下这是否应该和 OpenCompass 一样将 hellaswag 作为单纯的多选任务进行评测？暂时不太明白为什么要作为基于 logits 的评测实现。

## 补充 / Supplementary

将 `hellaswag_adapter.py` 的评测修改为多选类型后进行评测，结果略低于 OpenCompass 后端测出的。

Qwen2.5-1.5B-Instruct 评测结果：
- 修改后原生后端：56.28 acc
- OpenCompass 后端：61.74 acc

具体修改：
```python
# Copyright (c) Alibaba, Inc. and its affiliates.
import numpy as np
import os
import re

from evalscope.benchmarks import Benchmark, DataAdapter
from evalscope.constants import EvalType, OutputType
from evalscope.metrics import exact_match
from evalscope.utils.io_utils import jsonl_to_list
from evalscope.utils.logger import get_logger
from evalscope.utils.utils import ResponseParser

# flake8: noqa

logger = get_logger()


@Benchmark.register(
    name='hellaswag',
    pretty_name='HellaSwag',
    dataset_id='modelscope/hellaswag',
    model_adapter=OutputType.MULTIPLE_CHOICE,
    output_types=[OutputType.MULTIPLE_CHOICE, OutputType.GENERATION],
    subset_list=['default'],
    metric_list=['AverageAccuracy'],
    few_shot_num=0,
    train_split='train',
    eval_split='validation',
    prompt_template='{query}',  # noqa: E501
)
class HellaSwagAdapter(DataAdapter):

    def __init__(self, **kwargs):

        few_shot_num = kwargs.get('few_shot_num', 0)
        if few_shot_num != 0:
            logger.warning(f'few_shot_num should be 0 for HellaSwag, but got {few_shot_num}. Use 0-shot by default.')
            kwargs['few_shot_num'] = 0

        super().__init__(**kwargs)
        self.choices = ['0', '1', '2', '3']

    def load_from_disk(self, dataset_name_or_path, subset_list, work_dir, **kwargs) -> dict:
        data_dict = {}
        for subset_name in subset_list:
            data_dict[subset_name] = {}
            for split in [self.train_split, self.eval_split]:
                if os.path.exists(dataset_name_or_path):
                    file_path = os.path.join(dataset_name_or_path, f'hellaswag_{split}.jsonl')
                else:
                    file_path = os.path.join(work_dir, dataset_name_or_path, f'hellaswag_{split}.jsonl')
                if os.path.exists(file_path):
                    data_dict[subset_name][split] = jsonl_to_list(file_path)

        return data_dict

    def gen_prompt(self, input_d: dict, few_shot_list: list, **kwargs) -> dict:
        """
        Generate model prompt from raw data, unify the prompt format for HellaSwag benchmark.

        Args:
            input_d (dict): The raw input. A single data format of the HellaSwag:

            {
                'ind': 4,
                'activity_label': 'Removing ice from car',
                'ctx_a': 'Then, the man writes over the snow covering the window of a car, and a woman wearing winter clothes smiles.',
                'ctx_b': 'then',
                'ctx': 'Then, the man writes over the snow covering the window of a car, and a woman wearing winter clothes smiles. then',
                'endings': [', the man adds wax to the windshield and cuts it.', ', a person board a ski lift, while two men supporting the head of the person wearing winter clothes snow as the we girls sled.', ', the man puts on a christmas coat, knitted with netting.', ', the man continues removing the snow on his car.'],
                'source_id': 'activitynet~v_-1IBHYS3L-Y',
                'split': 'train',
                'split_type': 'indomain',
                'label': '3'
            }

        Returns:
            Refer to function: evalscope.benchmarks.data_adapter.DataAdapter.gen_prompt for details.
        """

        endings: list = [self._preprocess(ending) for ending in input_d['endings']]

        few_shot_prompts = [
            self._generate_prompt(input_d=sample, endings=endings, include_answer=True) for sample in few_shot_list
        ]
        context: str = '\n'.join(few_shot_prompts) + '\n'
        query = context.strip() + self._generate_prompt(input_d=input_d, endings=endings, include_answer=False)

        full_prompt = self.prompt_template.format(query=query)
        return self.gen_prompt_data(full_prompt)

    def get_gold_answer(self, input_d: dict) -> str:
        # Get the gold choice
        return input_d['label']

    def parse_pred_result(self, result: list, raw_input_d: dict = None, eval_type: str = EvalType.CHECKPOINT) -> str:
        """
        Parse the model output to get the answer. Could be the best choice index.

        Args:
            result: Predicted answer from the model. Usually a string for chat.
            raw_input_d: The raw input dict.
            eval_type: The evaluation type. e.g. checkpoint, service, custom.

        Returns:
            The parsed answer. Depending on the dataset. Usually a string for chat.
        """
        if self.model_adapter == OutputType.MULTIPLE_CHOICE:
            return result
        else:
            return ResponseParser.parse_first_option_with_choices(result, self.choices)

    def match(self, gold: str, pred: str) -> float:
        return exact_match(gold=str(gold), pred=str(pred))

    @classmethod
    def _preprocess(cls, text):
        text = text.strip()
        text = text.replace(' [title]', '. ')
        text = re.sub('\\[.*?\\]', '', text)
        text = text.replace('  ', ' ')
        return text

    @classmethod
    def _generate_prompt(cls, input_d: dict, endings: list, include_answer=True) -> str:
        """
        Generate prompt for HellaSwag dataset.

        Args:
            input_d: a single data of the hellaswag.
            endings:  preprocessed endings
            include_answer: bool

        Returns:

        """

        ctx = input_d['ctx_a'] + ' ' + input_d['ctx_b'].capitalize()
        # example: str = cls._preprocess(input_d['activity_label'] + ': ' + ctx)
        example: str = cls._preprocess(ctx)

        example += "\nQuestion: Which ending makes the most sense?"
        for i, ending in enumerate(endings):
            example += f'\n{i}. {ending}'
        example += "\nYou may choose from 0, 1, 2, 3.\nAnswer:"

        if include_answer:
            example += '{}\n\n'.format(endings[int(input_d['label'])])

        return example

```

修改后传给接口的输入：
```
<|im_start|>system
You are Qwen, created by Alibaba Cloud. You are a helpful assistant.<|im_end|>
<|im_start|>user\
How to strop a straight razor. Acquire a strop. These are relatively simple to get. You can get cheaper models from general stores online.
Question: Which ending makes the most sense?
0. These might cost you anything from $5 to $25. If you've been using a straight razor for a while, you'll want to invest in a strop that will last longer.
1. They are designed to be effective for a range of purposes except shaving. They are ideal for when you want to really imagine shaving off the hair.
2. Here's what one could buy : A strop is simply to simply hold one blade upright in one hand. Flat, sharp edges attach the blade to the body of the razor.
3. Look for one that has crimp at the ends. Crimp is a soft, feminine material.\n
You may choose from 0, 1, 2, 3.
Answer:<|im_end|>
<|im_start|>assistant
```

### LINKED ISSUE #382: [bug]: function `run_perf_benchmark`  can't work in second iteration.
## 问题描述 / Issue Description

The Python script using `run_perf_benchmark` gets stuck during the second loop. 
在自己脚本中使用函数run_perf_benchmark，在第二次循环的时候，函数run_perf_benchmark会卡住

## 使用的工具 / Tools Used
- [ ] Native / 原生框架
- [ ] Opencompass backend
- [ ] VLMEvalKit backend
- [ ] RAGEval backend
- [x] Perf / 模型推理压测工具
- [ ] Arena  /竞技场模式

## 执行的代码或指令 / Code or Commands Executed

```python

from evalscope.perf.main import run_perf_benchmark

for x_param in some_iter:
    run_perf_benchmark(x_param)

```

This piece of code will get stuck during the second iteration because of the following code.
[evalscope/perf/benchmark.py#L24](https://github.com/modelscope/evalscope/blob/main/evalscope/perf/benchmark.py#L24)
```python
# perf/benchmark.py : line 14.
# module's global variables
query_send_completed_event = asyncio.Event()
data_process_completed_event = asyncio.Event()
```
these  “asyncio events” would be initialized once during the script.  

The `statistic_benchmark_metric_worker` will complete early without any calls to `put()` and `task_done()` on `benchmark_data_queue`. After that, `benchmark_data_queue` receives some tasks. Consequently, the program gets stuck at `await benchmark_data_queue.join()`.

## Mitigation

**Modify source code**

My temporary approach is to add the event reset at the beginning of  func `benchmark`.
[evalscope/perf/benchmark.py#L21](https://github.com/modelscope/evalscope/blob/main/evalscope/perf/benchmark.py#L214)
```python
# perf/benchmark.py
@exception_handler
async def benchmark(args: Arguments) -> None:
    ...
    # line 214
    query_send_completed_event.clear()
    data_process_completed_event.clear() # U can only add this line to solve the problem
    ... 

``` 

**or Non-invasive way**

my python script
```python
from evalscope.perf.main import run_perf_benchmark
from evalscope.perf import benchmark

for x_param in some_iter:
    benchmark.query_send_completed_event.clear()
    benchmark.data_process_completed_event.clear() # U can only add this line to solve the problem

    run_perf_benchmark(x_param)

```

考虑到此项目的目的和特殊性，这里的做法仅仅适用与我的场景。是否定义此为bug并修复，还需要项目方综合考虑。

### PR DESCRIPTION
**Pull request by Yunnglin**
fix #381 
fix #382

### CODE REVIEW

File: docs/en/get_started/basic_usage.md [hunk 1]
```diff
@@ -188,6 +188,41 @@ evalscope eval \
...
+    api_key=os.getenv('DASHSCOPE_API_KEY'),
+    eval_type=EvalType.SERVICE,
+    datasets=[
+        # 'simple_qa',
+        'chinese_simpleqa',
+    ],
+    eval_batch_size=5,
+    limit=5,
+    judge_strategy=JudgeStrategy.DEFAULT,
+    judge_model_args={
```
Comment (wangxingjun778) [inline docs/en/get_started/basic_usage.md:211]: judge_model_args中的model_id、api_url跟上述TaskConfig中的model、api_url的关系是？ 看是否可以精简

File: docs/zh/get_started/basic_usage.md [hunk 2]
```diff
@@ -185,6 +185,39 @@ evalscope eval \
...
+    api_key= os.getenv('DASHSCOPE_API_KEY'),
+    eval_type=EvalType.SERVICE,
+    datasets=[
+        # 'simple_qa',
+        'chinese_simpleqa',
+    ],
+    eval_batch_size=5,
+    limit=5,
+    judge_strategy=JudgeStrategy.DEFAULT,
+    judge_model_args={
```
Comment (wangxingjun778) [inline docs/zh/get_started/basic_usage.md:208]: 同上

File: docs/en/get_started/basic_usage.md [hunk 3]
```diff
@@ -188,6 +188,41 @@ evalscope eval \
...
+    api_url='https://dashscope.aliyuncs.com/compatible-mode/v1',
+    api_key=os.getenv('DASHSCOPE_API_KEY'),
+    eval_type=EvalType.SERVICE,
+    datasets=[
+        # 'simple_qa',
+        'chinese_simpleqa',
+    ],
+    eval_batch_size=5,
+    limit=5,
+    judge_strategy=JudgeStrategy.DEFAULT,
```
Comment (wangxingjun778) [inline docs/en/get_started/basic_usage.md]: JudgeStrategy.DEFAULT 如果表示根据dataset配置来自适应定义的话，改成 JudgeStrategy.AUTO (同时value也改成 "auto" 是否更直观一些)

File: docs/en/get_started/basic_usage.md (hunk 1 above)
Comment (Yunnglin) [reply docs/en/get_started/basic_usage.md:211]: judge_model_args中的model_id、api_url 是 单独的，跟上述TaskConfig中的model、api_url没有联系，用户可以指定两个不同的服务

File: docs/en/get_started/basic_usage.md (hunk 3 above)
Comment (Yunnglin) [reply docs/en/get_started/basic_usage.md]: fixed

File: docs/en/get_started/basic_usage.md (hunk 1 above)
Comment (wangxingjun778) [reply docs/en/get_started/basic_usage.md:211]: 没有关联的话，可以将上下两处的model设置成不同的model_id，否则容易引起误解。
